linera-version/src/version_info.rs

```diff
@@ -0,0 +1,223 @@
+pub struct CrateVersion {
+    pub major: u64,
+    pub minor: u64,
+    pub patch: u64,
+}
+
+#[cfg(linera_version_building)]
+async_graphql::scalar!(
+    CrateVersion,
+    "CrateVersion",
+    "The version of the Linera crates used in this build"
+);
+
+impl From<semver::Version> for CrateVersion {
+    fn from(
+        semver::Version {
+            major,
+            minor,
+            patch,
+            ..
+        }: semver::Version,
+    ) -> Self {
+        Self {
+            major,
+            minor,
+            patch,
+        }
+    }
+}
+
+impl std::fmt::Display for CrateVersion {
+    fn fmt(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(formatter, "{}.{}.{}", self.major, self.minor, self.patch)
+    }
+}
+
+pub type Hash = std::borrow::Cow<'static, str>;
+
+#[cfg_attr(
+    linera_version_building,
+    derive(async_graphql::SimpleObject, serde::Deserialize, serde::Serialize)
+)]
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+/// The version info of a build of Linera.
+pub struct VersionInfo {
+    /// The crate version
+    pub crate_version: CrateVersion,
+    /// The git commit hash
+    pub git_commit: Hash,
+    /// Whether the git checkout was dirty
+    pub git_dirty: bool,
+    /// A hash of the RPC API
+    pub rpc_hash: Hash,
+    /// A hash of the GraphQL API
+    pub graphql_hash: Hash,
+    /// A hash of the WIT API
+    pub wit_hash: Hash,
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    #[error("failed to interpret cargo-metadata: {0}")]
+    CargoMetadata(#[from] cargo_metadata::Error),
+    #[error("no such package: {0}")]
+    NoSuchPackage(String),
+    #[error("I/O error: {0}")]
+    Io(#[from] std::io::Error),
+    #[error("glob error: {0}")]
+    Glob(#[from] glob::GlobError),
+    #[error("pattern error: {0}")]
+    Pattern(#[from] glob::PatternError),
+}
+
+struct Outcome {
+    status: std::process::ExitStatus,
+    output: String,
+}
+
+pub type Result<T> = std::result::Result<T, Error>;
+
+fn get_hash(
+    relevant_paths: &mut Vec<PathBuf>,
+    metadata: &cargo_metadata::Metadata,
+    package: &str,
+    glob: &str,
+) -> Result<String> {
+    use base64::engine::{general_purpose::STANDARD_NO_PAD, Engine as _};
+    use sha3::Digest as _;
+
+    let Some(package_root) = get_package_root(metadata, package) else {
+        return Ok("package not used".to_owned());
+    };
+
+    let mut hasher = sha3::Sha3_256::new();
+    let mut buffer = [0u8; 4096];
+
+    let package_glob = format!("{}/{}", package_root.display(), glob);
+
+    for path in glob::glob(&package_glob)? {
+        let path = path?;
+        let mut file = std::fs::File::open(&path)?;
+        relevant_paths.push(path);
+        while file.read(&mut buffer)? != 0 {
+            hasher.update(buffer);
+        }
+    }
+
+    Ok(STANDARD_NO_PAD.encode(hasher.finalize()))
+}
+
+fn run<'a>(cmd: &str, args: &[&str], stdin: impl Into<Option<&'a str>>) -> Result<Outcome> {
```

Comment:
nit: There was a debate in the team and we tend to not use `impl Into<Option<..>>` any more without a good reason. (I can't see where a non-empty stdin arg is passed btw. Was this meant for testing then forgotten?)

Reply:
Oh, no, I think it's just because I used to use `sha256sum`.  Thanks for the catch :)